src/Take.Blip.Client/Receivers/ContactMessageReceiverBase.cs

```diff
@@ -20,62 +20,76 @@ public abstract class ContactMessageReceiverBase : IMessageReceiver, IDisposable
-            
-            if (contact == null)
+            try
             {
-                try
+                using (var cts = new CancellationTokenSource(_getContactTimeout))
                 {
-                    // Second, try from the roster.
-                    contact = await _contactExtension.GetAsync(identity, cancellationToken);
-                }
-                catch (LimeException ex) when (ex.Reason.Code == ReasonCodes.COMMAND_RESOURCE_NOT_FOUND || ex.Reason.Code == ReasonCodes.COMMAND_RESOURCE_NOT_SUPPORTED) { }
+                    // First, tries get it from the cache, if configured.
+                    if (_cacheLocally)
+                    {
+                        contact = _contactCache.Get(identity.ToString()) as Contact;
+                    }
 
-                // Third, try from the directory.
-                if (contact == null)
-                {
-                    try
+                    if (contact == null)
                     {
-                        contact = await GetContactFromDirectoryAsync(identity, cancellationToken);
+                        try
+                        {
+                            // Second, try from the roster.
+                            contact = await _contactExtension.GetAsync(identity, cts.Token);
+                        }
+                        catch (LimeException ex) when (ex.Reason.Code == ReasonCodes.COMMAND_RESOURCE_NOT_FOUND || ex.Reason.Code == ReasonCodes.COMMAND_RESOURCE_NOT_SUPPORTED) { }
+
+                        // Third, try from the directory.
+                        if (contact == null)
+                        {
+                            try
+                            {
+                                contact = await GetContactFromDirectoryAsync(identity, cts.Token);
+                            }
+                            catch (LimeException ex) when (ex.Reason.Code == ReasonCodes.COMMAND_RESOURCE_NOT_FOUND || ex.Reason.Code == ReasonCodes.COMMAND_RESOURCE_NOT_SUPPORTED) { }
+                        }
+
+                        // Stores in the cache, if configured.
+                        if (contact != null && _cacheLocally)
+                        {
+                            _contactCache.Set(identity.ToString(), contact, DateTimeOffset.UtcNow.Add(_cacheExpiration));
+                        }
                     }
-                    catch (LimeException ex) when (ex.Reason.Code == ReasonCodes.COMMAND_RESOURCE_NOT_FOUND || ex.Reason.Code == ReasonCodes.COMMAND_RESOURCE_NOT_SUPPORTED) { }
-                }
-                
-                // Stores in the cache, if configured.
-                if (contact != null && _cacheLocally)
-                {
-                    _contactCache.Set(identity.ToString(), contact, DateTimeOffset.UtcNow.Add(_cacheExpiration));
                 }
             }
-
-            await ReceiveAsync(envelope, contact, cancellationToken);
+            finally
```

Comment:
This is dangerous because it can be called in a context that an exception is thrown by this method, leading to unknown behavior from the caller class. Besides that, if the cancellationToken is cancelled, in this call the cancellationToken will be invalid.

It is better to remove the `finally` clause and add catch clause above handling `OperationCanceledException` only when the contact timeout is reached (`when (cts.IsCancelledRequested)`) to allow the process continue without the value of the contact.